---
title: "Cleaned Manuscript Code - Pitcher Field Exp"
author: "Aldo A. Arellano"
date: "8/15/2023"
output: html_document
---
```{r}
library(ape)
library(magrittr)
library(data.table)
library(phytools)
library(xfun)
library(decontam)
library(phyloseq)
library(ggplot2)
library(breakaway)
library(DivNet)
library(dplyr)
library(cowplot)
library(wesanderson)
library(RColorBrewer)
library(beanplot)
library(vegan)
library(ranacapa)
library(ggpubr)
library(gridExtra)
library(reshape)
library(grid)
library(viridis)
library(pander)
library(gplots)
library(VennDiagram)
library(ALDEx2)
library(tidyverse)
library(pander)
library(breakaway)
library(DivNet)
library(purrr)
library(FSA)
library(pairwiseAdonis)
library(RLRsim)
library(clusterSim)
library(cluster)
library(ade4)
library(BiotypeR)
library(tibble)
```

Read in the phyloseq object for analysis, for more information on decontamination procedure see:https://github.com/benjjneb/decontam 
```{r}
setwd("[...]/github pitcher plant addition ms resources")
phy_object_final<-readRDS("final_pitcher_phyloseq")
```

*Patterns of mosquito colonization and treatment effects in field manipulations over time*

Were manipulations treatments enough to influence number of larvae in pitchers over the course of the field experiment?
```{r}
setwd("[...]/github pitcher plant addition ms resources")
meta_csv<-read.csv("metadata_pitch.csv")
meta_csv<-meta_csv %>% drop_na(Sample)

meta_0<-meta_csv[meta_csv$Day.Number==0,]
meta_11<-meta_csv[meta_csv$Day.Number==11,]
meta_21<-meta_csv[meta_csv$Day.Number==21,]
meta_32<-meta_csv[meta_csv$Day.Number==32,]
meta_46<-meta_csv[meta_csv$Day.Number==46,]
meta_56<-meta_csv[meta_csv$Day.Number==56,]
meta_67<-meta_csv[meta_csv$Day.Number==67,]

kruskal.test(Total_Resident~Treatment,data=meta_0)
kruskal.test(Total_Resident~Treatment,data=meta_11)
kruskal.test(Total_Resident~Treatment,data=meta_21)
kruskal.test(Total_Resident~Treatment,data=meta_32)
kruskal.test(Total_Resident~Treatment,data=meta_46)
kruskal.test(Total_Resident~Treatment,data=meta_56)
kruskal.test(Total_Resident~Treatment,data=meta_67)

dunnTest(Total_Resident~Treatment,data=meta_0, method="bh")
dunnTest(Total_Resident~Treatment,data=meta_11, method="bh")
dunnTest(Total_Resident~Treatment,data=meta_21, method="bh")
dunnTest(Total_Resident~Treatment,data=meta_32, method="bh")
dunnTest(Total_Resident~Treatment,data=meta_46, method="bh")
dunnTest(Total_Resident~Treatment,data=meta_56, method="bh")
dunnTest(Total_Resident~Treatment,data=meta_67, method="bh")
```

Conducting PERMANOVA and PERMDISP: full set
```{r}
###full set
ps_clr<-microbiome::transform(phy_object_final, "clr") 

clr.dist <- dist(t(otu_table(ps_clr)), method="euclidean")

meta<-sample_data(ps_clr)
meta<-cbind(rownames(meta), data.frame(meta, row.names=NULL))
colnames(meta)[1] <- "Sample ID"

clr_df_dist<-as.matrix(clr.dist)
clr_df_dist<-as.data.frame(clr_df_dist)
clr_df_dist <- cbind(rownames(clr_df_dist), data.frame(clr_df_dist, row.names=NULL))
colnames(clr_df_dist)[1] <- "Sample ID"

joined<-inner_join(clr_df_dist,meta, by = "Sample ID")

set.seed(1)
ps.disper_sample_type_clr <- betadisper(clr.dist, joined$Start_End)
disper_test_sample_type_clr<-permutest(ps.disper_sample_type_clr, pairwise = TRUE, p.adjust.methods = "bonferroni")
summ<-summary(permustats(disper_test_sample_type_clr))
# summ
a<-purrr::pluck(summ,'statistic')
b<-purrr::pluck(summ,'p')
c<-as.data.frame(a)
d<-as.data.frame(b)
rownames(d) = rownames(c)
# rownames(c)
d <- cbind(rownames(d), data.frame(d, row.names=NULL))
c <- cbind(rownames(c), data.frame(c, row.names=NULL))
colnames(c)= c("Comparison","PERMDISP Test Statistic")
colnames(d)= c("Comparison","PERMDISP p-value")
FIELD_clr_betadisper_table<-merge(c,d)

adonis<-adonis2(clr.dist~Type,data=joined, permutations = 999, pairwise=TRUE)
```

Conducting PERMANOVA and PERMDISP: start set
```{r}
#start only
start_phy = prune_samples(sample_data(phy_object_final)$Type == "Start", phy_object_final) 
start_phy = prune_samples(sample_data(start_phy)$Treatment == "Addition"|sample_data(start_phy)$Treatment == "Variable" | sample_data(start_phy)$Treatment == "Removal",start_phy) 
start_phy = prune_taxa(taxa_sums(start_phy) > 0, start_phy)
ps_clr<-microbiome::transform(start_phy, "clr") 

clr.dist <- dist(t(otu_table(ps_clr)), method="euclidean")

meta<-sample_data(ps_clr)
meta<-cbind(rownames(meta), data.frame(meta, row.names=NULL))
colnames(meta)[1] <- "Sample ID"

clr_df_dist<-as.matrix(clr.dist)
clr_df_dist<-as.data.frame(clr_df_dist)
clr_df_dist <- cbind(rownames(clr_df_dist), data.frame(clr_df_dist, row.names=NULL))
colnames(clr_df_dist)[1] <- "Sample ID"

joined<-inner_join(clr_df_dist,meta, by = "Sample ID")

set.seed(1)
ps.disper_sample_type_clr <- betadisper(clr.dist, joined$Treatment)
disper_test_sample_type_clr<-permutest(ps.disper_sample_type_clr, pairwise = TRUE, p.adjust.methods = "bonferroni")

summary(permustats(disper_test_sample_type_clr))
adonis2(clr.dist~Treatment,data=joined, permutations = 999)
```

Conducting PERMANOVA and PERMDISP: harvest set
```{r}
#harvest only
harvest_phy= prune_samples(sample_data(phy_object_final)$Type == "Harvest", phy_object_final) 
harvest_phy = prune_samples(sample_data(harvest_phy)$Treatment == "Addition"|sample_data(harvest_phy)$Treatment == "Variable" | sample_data(harvest_phy)$Treatment == "Removal",harvest_phy) 
harvest_phy = prune_taxa(taxa_sums(harvest_phy) > 0, harvest_phy)
ps_clr<-microbiome::transform(harvest_phy, "clr") 

clr.dist <- dist(t(otu_table(ps_clr)), method="euclidean")

meta<-sample_data(ps_clr)
meta<-cbind(rownames(meta), data.frame(meta, row.names=NULL))
colnames(meta)[1] <- "Sample ID"

clr_df_dist<-as.matrix(clr.dist)
clr_df_dist<-as.data.frame(clr_df_dist)
clr_df_dist <- cbind(rownames(clr_df_dist), data.frame(clr_df_dist, row.names=NULL))
colnames(clr_df_dist)[1] <- "Sample ID"

joined<-inner_join(clr_df_dist,meta, by = "Sample ID")

set.seed(1)
ps.disper_sample_type_clr <- betadisper(clr.dist, joined$Treatment)
disper_test_sample_type_clr<-permutest(ps.disper_sample_type_clr, pairwise = TRUE)
summ<-summary(permustats(disper_test_sample_type_clr))
p.adjust(summ$p,method = "bonferroni", n=4)

set.seed(1)
adonis2(clr.dist~Treatment,data=joined, permutations = 999)
pairwise.adonis(clr.dist,joined$Treatment, p.adjust.m = "bonferroni", reduce = NULL)
```

Differential abundance by timepoint  (sample code):
```{r}
# aldex.clr.full<-aldex.clr(as.data.frame(otu_table(new_phy)),
#                              sample_data(new_phy)$Type,
#                              mc.samples=128,
#                              denom="all",
#                              verbose=TRUE,
#                              useMC=FALSE)
# 
# full_set_aldex_t_test_asv<-aldex.ttest(aldex.clr.full, paired.test = FALSE, hist.plot = FALSE, verbose = FALSE)
# full_set_aldex.effect_asv <- aldex.effect(aldex.clr.full, CI=T, verbose=FALSE)
# 
# full_set_aldex.all.df_asv <- data.frame(full_set_aldex_t_test_asv,full_set_aldex.effect_asv)
# 
# par(mfrow=c(1,2))
# aldex.plot(full_set_aldex.all.df_asv, type="MA", test="welch")
# aldex.plot(full_set_aldex.all.df_asv, type="MW", test="welch")

# taxa_info <- data.frame(tax_table(new_phy))
# taxa_info <- taxa_info %>% rownames_to_column(var = "ASV")
# 
# sig_aldex2 <- full_set_aldex.all.df_asv %>%
#   rownames_to_column(var = "ASV") %>%
#   filter(we.eBH < 0.05) %>%
#   arrange(we.eBH) %>%
#   dplyr::select(ASV, we.eBH, wi.eBH, diff.btw)
# sig_aldex2 <- left_join(sig_aldex2, taxa_info)

# write.csv(sig_aldex2,"diff taxa full start vs harvest.csv")
```

Differential abundance testing at start or harvest by treatment (sample code):
```{r}
## sample code, but repeated for every taxonomic rank and for each timepoint individually:

# harvest_phy = prune_samples(sample_data(new_phy)$Type == "Harvest", new_phy)
# 
# harvest_nonzero<-prune_taxa(taxa_sums(harvest_phy) > 0, harvest_phy)
#  harvest_nonzero<-tax_glom(harvest_nonzero, taxrank="Family", NArm=TRUE, bad_empty=c(NA, "", " ", "\t"))
# 
# aldex.clr.full<-aldex.clr(as.data.frame(otu_table(harvest_nonzero)),
#                              sample_data(harvest_nonzero)$Treatment,
#                              mc.samples=128,
#                              denom="all",
#                              verbose=TRUE,
#                              useMC=FALSE)
# 
# glm.kru.harvest.data.cluster<-aldex.kw(aldex.clr.full)
# 
# taxa_info <- data.frame(tax_table(harvest_nonzero))
# taxa_info <- taxa_info %>% rownames_to_column(var = "ASV")
# 
# sig_aldex2 <- glm.kru.harvest.data.cluster %>%
#   rownames_to_column(var = "ASV") %>%
#   filter(kw.eBH < 0.05) %>%
#   arrange(kw.eBH) %>%
#   dplyr::select(ASV, glm.eBH, kw.eBH)
# sig_aldex2 <- left_join(sig_aldex2, taxa_info)
```

Generating sample summary dataframe, including diversity metrics: First diversity estimation using DivNet and breakaway (sample code)
```{r}
#Shannon
# DivNet can take a long time to run!
# shannon<-DivNet::divnet(phy_object_final, base="e5e45fe5960b838c45765068621efcd5", ncores = 6)
# 
# shannon_est <- shannon$shannon %>% summary %$% estimate
# shannon_ses <- sqrt(shannon$`shannon-variance`)
# 
# df<-as.data.frame(shannon_est)
# colnames(df)="Shannon Sample-Wise"
# shannon_final<- tibble::rownames_to_column(df, "Sample ID")
# 
# df<-as.data.frame(shannon_ses)
# colnames(df)="Shannon Error"
# shannon_error <- tibble::rownames_to_column(df, "Sample ID")
# 
# shannon_merge<-merge(shannon_final, shannon_error)
# 
# # rich
# field_richness<-breakaway::breakaway(phy_object_final)
# 
# rich<-field_richness %>% summary %$% estimate
# rich_error<-field_richness %>% summary %$% error
# 
# df<-as.data.frame(rich)
# colnames(df)="Breakaway Estimate of Richness"
# rich_final <- tibble::rownames_to_column(df, "Sample ID")
# 
# df<-as.data.frame(rich_error)
# colnames(df)="Breakaway Error"
# rich_error_final <- tibble::rownames_to_column(df, "Sample ID")
# 
# rich_merge<-merge(rich_final,rich_error_final)
# 
# div_measures_field<-merge(shannon_merge,rich_merge)
#saveRDS(div_measures_field,"field diversity sample-wise")

## Resultant dataframe:
setwd("[...]/github pitcher plant addition ms resources")
shannon<-readRDS("field diversity sample-wise")
```

Next, grab some summary data for sequencing depth and ASV counts from phyloseq object:
```{r}
 #Count of ASV richness
 ASV_Richness<-colSums(otu_table(phy_object_final)!=0)

 #Total reads per sample
 Total_Reads<-sample_sums(phy_object_final)

 #some temporary objects to get things wrangled and labeled:
 a<-as.data.frame(ASV_Richness)
 a <- tibble::rownames_to_column(a, "Sample ID")
 b<-as.data.frame(Total_Reads)
 b <- tibble::rownames_to_column(b, "Sample ID")
 new<-merge(a,b)

 a<-as.data.frame(sample_data(phy_object_final))
 b<-as.matrix(a)
 c<-as.data.frame(b)
 df <- tibble::rownames_to_column(c, "Sample ID")

 merged_sample_ID<-merge(df,shannon)
 merged_sample_ID<-merge(merged_sample_ID,new)

 # saveRDS(merged_sample_ID,"betta_test_df")
```

And finally conduct hypothesis testing by time and treatment group:
```{r}
setwd("[...]/github pitcher plant addition ms resources")
betta_test_df<-readRDS("betta_test_df")
betta_test_df<-betta_test_df[betta_test_df$Type == 'Start'|betta_test_df$Type == 'Harvest' ,]

#time first
set.seed(123)
a <- betta(formula = `Shannon Sample-Wise` ~ Type, 
                      ses = `Shannon Error`, data = betta_test_df)
a$table

b <- betta(formula = `Breakaway Estimate of Richness` ~ Type, 
                      ses = `Breakaway Error`, data = betta_test_df)
b$table

#now looking at just the starting samples

betta_test_df_start<-betta_test_df[betta_test_df$Type=="Start",]
betta_test_df_start<-betta_test_df_start %>% drop_na(Treatment)
betta_test_df_start$Treatment<-factor(betta_test_df_start$Treatment, levels=c("Variable","Removal", "Addition"))
set.seed(123)
c <- betta(formula = `Shannon Sample-Wise` ~ Treatment, 
                      ses = `Shannon Error`, data = betta_test_df_start)
c$table

d <- betta(formula = `Breakaway Estimate of Richness` ~ Treatment, 
                      ses = `Breakaway Error`, data = betta_test_df_start)
d$table

set.seed(123)
betta_test_df_start$Treatment<-factor(betta_test_df_start$Treatment, levels=c("Removal", "Addition"))
betta_test_df_start<-betta_test_df_start %>% drop_na(Treatment)

e <-betta(formula = `Shannon Sample-Wise` ~ Treatment, 
                      ses = `Shannon Error`, data = betta_test_df_start)
e$table

f <- betta(formula = `Breakaway Estimate of Richness` ~ Treatment, 
                      ses = `Breakaway Error`, data = betta_test_df_start)
f$table

#now looking at just the harvest samples

betta_test_df_harvest<-betta_test_df[betta_test_df$Type=="Harvest",]
betta_test_df_harvest<-betta_test_df_harvest %>% drop_na(Treatment)
betta_test_df_harvest$Treatment<-factor(betta_test_df_harvest$Treatment, levels=c("Variable","Removal","Addition"))
set.seed(123)
g <- betta(formula = `Shannon Sample-Wise` ~ Treatment, 
                      ses = `Shannon Error`, data = betta_test_df_harvest)
g$table

h <- betta(formula = `Breakaway Estimate of Richness` ~ Treatment, 
                      ses = `Breakaway Error`, data = betta_test_df_harvest)
h$table

set.seed(123)
betta_test_df_harvest$Treatment<-factor(betta_test_df_harvest$Treatment, levels=c("Removal", "Addition"))
betta_test_df_harvest<-betta_test_df_harvest %>% drop_na(Treatment)

i<- betta(formula = `Breakaway Estimate of Richness` ~ Treatment, 
                      ses = `Breakaway Error`, data = betta_test_df_harvest)
i$table

j <- betta(formula = `Breakaway Estimate of Richness` ~ Treatment, 
                      ses = `Breakaway Error`, data = betta_test_df_harvest)
j$table
```

*Several bacterial ASVs correlate with mosquito larval colonization*

```{r}
setwd("[...]/github pitcher plant addition ms resources")
phy_object_final<-readRDS("final_pitcher_phyloseq")

#remove samples that did not end up with treatment group information
phy_object_final<-prune_samples(sample_data(phy_object_final)$Treatment!= "NA", phy_object_final)

# input is CLR for continuous. aldex() for categorical does clr automatically
aldex2_pitcher_clr <- aldex.clr(as.data.frame(otu_table(phy_object_final)),
                             sample_data(phy_object_final)$Total_Resident,
                             mc.samples=128,
                             denom="all",
                             verbose=TRUE,
                             useMC=FALSE)

# aldex2_pitcher_corr <- aldex.corr(aldex2_pitcher_clr, sample_data(phy_object_final)$Total_Resident)
# aldex2_pitcher_corr <- aldex2_pitcher_corr %>% rownames_to_column(var = "ASV")
aldex2_pitcher_corr_significant<- aldex2_pitcher_corr%>% filter(pearson.eBH<0.05) 
aldex2_pitcher_corr_significant<-aldex2_pitcher_corr_significant[,c(1:4)]


taxa_info <- data.frame(tax_table(phy_object_final))
taxa_info <- taxa_info %>% rownames_to_column(var = "ASV")

#then join to taxonomy by ASV name and generate final data frame of correlated asvs with taxonomy
aldex2_corr_asv <- merge(aldex2_pitcher_corr_significant, taxa_info, by="ASV")

#negative
aldex2_corr_asv[aldex2_corr_asv$pearson.ecor<0,]
#positive
aldex2_corr_asv[aldex2_corr_asv$pearson.ecor>0,]

```

*Microbial community function is related to larval abundance* 

Testing for by-treatment significance (similar approach used for by-biotype analyses below)
```{r}
setwd("[...]/github pitcher plant addition ms resources")
df<-read.csv("larvae_nutrients_respiration.csv")
df<-df[,-c(1:2)]

df<-na.omit(df)
w<-NULL

for(i in 1:(ncol(df)-1)) { 
  blip<-aov(df[,i] ~ df$Treatment, data= df) 
  blip<-summary(blip)
  bloop<-(c(colnames(df)[i],blip[[1]][["F value"]][[1]],blip[[1]][["Df"]][[1]],blip[[1]][["Df"]][[2]],blip[[1]][["Pr(>F)"]][[1]]))
w<-rbind(w,bloop)
}

all_aov_results<-as.data.frame(w)
colnames(all_aov_results)<-c("Covariate","F-statistic","df1","df2","p_value")
all_aov_results<-data.frame(all_aov_results,row.names = NULL)
sig_aov_results<-all_aov_results %>% filter (p_value < 0.05)
```

Unsurprising that some larval variables are correlated with treatment groups, in addition the respiration variables end up not meeting assumptions of parametric stats so proceed with non-parametric approach
```{r}
kruskal.test(RespRate~Treatment,data=df)
dunnTest(RespRate ~ Treatment,data=df,method="bh")
```

Now for enzyme analysis which we conducted in tandem with several visitations not just at harvest.
Starting with protease activity encoding larval density as a treatment effect:
```{r}
setwd("[...]/github pitcher plant addition ms resources")
enzymes<-read.csv("all_field_entry.csv")

new_enzyme<-enzymes %>% filter(Treatment!="NA")
levels(new_enzyme$Treatment)
new_enzyme<-new_enzyme %>% filter(Corrected_Pro!="NA")

new_enzyme<-new_enzyme[,c(1,3:5,10,13:14)]
new_enzyme$Plant<-as.factor(new_enzyme$Plant)
new_enzyme$Treatment<-as.factor(new_enzyme$Treatment)
new_enzyme$log_Pro<-as.numeric(new_enzyme$log_Pro)
new_enzyme$log_Chit<-as.numeric(new_enzyme$log_Chit)
new_enzyme$Pitcher<-as.factor(new_enzyme$Pitcher)
pro_set<-new_enzyme[,-7]
pro_set$Treatment<-factor(pro_set$Treatment, levels=c("Variable","Removal","Addition"))

set.seed(1234)
mix.int.all <-lmerTest::lmer(log_Pro ~ relevel(Treatment,ref="Variable") + Exp_Day + (1|Plant/Pitcher), data = pro_set)
anova(mix.int.all)
summary(mix.int.all)

set.seed(1234)
mix.int.all_interaction <-lmerTest::lmer(log_Pro ~ relevel(Treatment,ref="Variable") * Exp_Day + (1|Plant/Pitcher), data = pro_set)
anova(mix.int.all_interaction)
summary(mix.int.all_interaction)

temp_pro<-interact_plot(mix.int.all_interaction, pred = Exp_Day, modx= Treatment,
              interval = TRUE,
              int.width = 0.8)

anova(mix.int.all,mix.int.all_interaction)

#function to summarize data
summary_se <- function(.data, measure_var, ..., .ci = 0.95, na.rm = FALSE) {

  measure_var <- dplyr::enquo(measure_var)
  group_var <- dplyr::enquos(...)

  .data %>%
    group_by(!!! group_var) %>%
    summarise(mean = mean(!! measure_var, na.rm = na.rm),
              sd = sd(!! measure_var, na.rm = na.rm),
              n = n(),
              se = sd/sqrt(n),
              ci = se * qt(.ci/2 + 0.5, n-1)) %>%
    ungroup()
}
summary_df<-summary_se(new_enzyme,log_Pro,Treatment,Exp_Day)

#base line plot for enzyme activity
ggplot(data=summary_df, aes(x=Exp_Day, y=mean, group=Treatment)) +
  geom_line(aes(color=Treatment),lwd=1) + geom_errorbar(aes(ymin=mean-se, ymax=mean+se, color= Treatment),lwd=1,position=position_jitter(width = 1, height = NULL, seed = NA), width=.2)+
  theme_cowplot() + xlab("Experiment Day") + ylab("log (Protease activity)") + theme(panel.grid.major = element_line(colour = "grey70", size = 0.1), panel.grid.minor = element_blank())

# Assessing random effect significance as detailed at: http://users.stat.umn.edu/~gary/book/RExamples/nested-and-mixed-models.html 

#make a couple more models to test significance of additional terms using nested models approach
mix.sub.1 <-lmerTest::lmer(log_Pro ~ relevel(Treatment,ref="Variable") + Exp_Day + (1|Plant:Pitcher), data = pro_set)
mix.sub.2 <-lmerTest::lmer(log_Pro ~ relevel(Treatment,ref="Variable") + Exp_Day + (1|Plant), data = pro_set)

#pitcher
set.seed(1234)
exactRLRT(mix.sub.1,mix.int.all,mix.sub.2)

#plant
set.seed(1234)
exactRLRT(mix.sub.2,mix.int.all,mix.sub.1)
```

Next up chitinase activity:
```{r}
chit_set<-new_enzyme[,-6]
chit_set$Treatment<-factor(chit_set$Treatment, levels=c("Variable","Removal","Addition"))

summary_df<-summary_se(new_enzyme,log_Chit,Treatment,Exp_Day)

#base line plot for enzyme activity
ggplot(data=summary_df, aes(x=Exp_Day, y=mean, group=Treatment)) +
  geom_line(aes(color=Treatment),lwd=1) + geom_errorbar(aes(ymin=mean-se, ymax=mean+se, color= Treatment),lwd=1,position=position_jitter(width = 1, height = NULL, seed = NA), width=.2)+
  theme_cowplot() + xlab("Experiment Day") + ylab("log (Chitinase activity)") + theme(panel.grid.major = element_line(colour = "grey70", size = 0.1), panel.grid.minor = element_blank())

set.seed(1234)
mix.int.all <-lmerTest::lmer(log_Chit ~ relevel(Treatment,ref="Variable") + Exp_Day + (1|Plant/Pitcher), data = chit_set)
anova(mix.int.all)
summary(mix.int.all)

set.seed(1234)
mix.int.all_interaction <-lmerTest::lmer(log_Chit ~ relevel(Treatment,ref="Variable") * Exp_Day + (1|Plant/Pitcher), data = chit_set)
anova(mix.int.all_interaction)
summary(mix.int.all_interaction)

temp_chit<-interact_plot(mix.int.all_interaction, pred = Exp_Day, modx= Treatment,
              interval = TRUE,
              int.width = 0.8)

anova(mix.int.all,mix.int.all_interaction)

# Assessing random effect significance as detailed at: http://users.stat.umn.edu/~gary/book/RExamples/nested-and-mixed-models.html 

#make a couple more models
mix.sub.1 <-lmerTest::lmer(log_Chit ~ relevel(Treatment,ref="Variable") + Exp_Day + (1|Plant:Pitcher), data = chit_set)
mix.sub.2 <-lmerTest::lmer(log_Chit ~ relevel(Treatment,ref="Variable") + Exp_Day + (1|Plant), data = chit_set)

#pitcher
set.seed(1234)
exactRLRT(mix.sub.1,mix.int.all,mix.sub.2)

#plant
set.seed(1234)
exactRLRT(mix.sub.2,mix.int.all,mix.sub.1)
```

What about when larval density is encoded as a continuous variable?
Protease first:
```{r}
set.seed(1234)
mix.int.all <-lmerTest::lmer(log_Pro ~ Total_Resident + Exp_Day + (1|Plant:Pitcher), data = pro_set)
mix.int.all_interact_continuous<-lmerTest::lmer(log_Pro ~ Total_Resident * Exp_Day + (1|Plant:Pitcher), data = pro_set)

anova(mix.int.all)
summary(mix.int.all)

anova(mix.int.all_interact_continuous)
summary(mix.int.all_interact_continuous)

# install.packages("interactions")
library("interactions")

protease<-interact_plot(mix.int.all_interact_continuous, pred = Exp_Day, modx= Total_Resident,
              interval = TRUE,
              int.width = 0.8)

protease + xlab("Experiment Day") + ylab("log(Protease activity) (nM AMC/min)") + labs(fill="Total resident")

anova(mix.int.all,mix.int.all_interact_continuous)

mix.sub.1 <-lmerTest::lmer(log_Pro ~Total_Resident * Exp_Day + (1|Plant:Pitcher), data = pro_set)

mix.sub.2 <-lmerTest::lmer(log_Pro ~ Total_Resident * Exp_Day + (1|Plant), data = pro_set)

#pitcher
set.seed(1234)
exactRLRT(mix.sub.1,mix.int.all_interact_continuous,mix.sub.2)

#plant
set.seed(1234)
exactRLRT(mix.sub.2,mix.int.all_interact_continuous,mix.sub.1)
```

Next chitinase:
```{r}
set.seed(1234)
mix.int.all <-lmerTest::lmer(log_Chit ~ Total_Resident + Exp_Day + (1|Plant/Pitcher), data = chit_set)

anova(mix.int.all)
summary(mix.int.all)

set.seed(1234)
mix.int.all_interact_continuous<-lmerTest::lmer(log_Chit ~ Total_Resident * Exp_Day + (1|Plant:Pitcher), data = chit_set)
anova(mix.int.all_interact_continuous)
summary(mix.int.all_interact_continuous)

anova(mix.int.all,mix.int.all_interact_continuous)

chitinase<-interact_plot(mix.int.all_interact_continuous, pred = Exp_Day, modx= Total_Resident,
              interval = TRUE,
              int.width = 0.8)

chitinase$labels$fill <- "Total resident"  

chitinase + xlab("Experiment Day") + ylab("log(Chitinase activity) (nM MU/min)") + labs(fill="Total resident")
  
mix.sub.1 <-lmerTest::lmer(log_Chit ~ Total_Resident + Exp_Day + (1|Plant:Pitcher), data = chit_set)

mix.sub.2 <-lmerTest::lmer(log_Chit ~ Total_Resident + Exp_Day + (1|Plant), data = chit_set)

#pitcher
set.seed(1234)
exactRLRT(mix.sub.1,mix.int.all,mix.sub.2)

#plant
set.seed(1234)
exactRLRT(mix.sub.2,mix.int.all,mix.sub.1)
```

*Independent clustering supports successional differences in composition and function associated with mosquito larval density*

PAM clustering (sample code):
```{r}
# full set (also repeated on start and harvest set)
# setwd('C:/Users/aldoa/OneDrive/Desktop/Pitcher_Exp_2020_seq/phyloseq/')
# read<-readRDS("final_pitcher_phyloseq")
# 
# genus_glom<-tax_glom(read, taxrank="Genus", NArm=TRUE, bad_empty=c(NA, "", " ", "\t"))
# ps_clr_genus<-microbiome::transform(genus_glom, "clr")
# otu_clr_genus<- as.data.frame(otu_table(ps_clr_genus))
# otu_genus<-t(otu_clr_genus)
# matrix_clr_dist_genus<-vegdist(otu_genus, method="euclidean", binary=FALSE, diag=FALSE, upper=FALSE,
#                       na.rm = TRUE)
# data=otu_genus
# data.dist=matrix_clr_dist_genus
# 
# pam.clustering=function(x,k) { # x is a distance matrix and k the number of clusters
#                          require(cluster)
#                          cluster = as.vector(pam(as.dist(x), k, diss=TRUE)$clustering)
#                          return(cluster)
# }
# 
# data.cluster=pam.clustering(data.dist, k=2)
# 
# nclusters=NULL
# 
# 	for (k in 1:20) { 
# 		if (k==1) {
# 			nclusters[k]=NA 
# 		} else {
# 			data.cluster_temp=pam.clustering(data.dist, k)
# 			nclusters[k]=index.G1((data),data.cluster_temp,  d = data.dist,
# 			centrotypes = "medoids")
# 		}
# 	}
# 	
# plot(nclusters, type="h", xlab="k clusters", ylab="CH index")
# mean(silhouette(data.cluster, data.dist)[,3])
```

3 x 3 contingency for biotyope membership correlation to treatment groups conducted online at: http://vassarstats.net/fisher3x3.html

Conservative consensus approach for identifying taxa enriched or depauperate in harvest timepoint biotypes - Beginning with conducting Between Class Analysis (BCA)
```{r}
setwd("[...]/github pitcher plant addition ms resources")
with_biotype_genus<-readRDS("with_biotype_genus_harvest")

ps_clr_harvest_genus<-microbiome::transform(with_biotype_genus, "clr")
otu_harvest_clr_genus<- as.data.frame(otu_table(ps_clr_harvest_genus))
otu_harvest_clr_genus<-t(otu_harvest_clr_genus)
matrix_harvest_clr_dist_genus<-vegdist(otu_harvest_clr_genus, method="euclidean", binary=FALSE, diag=FALSE, upper=FALSE, na.rm = TRUE)
data.dist=matrix_harvest_clr_dist_genus

harvest.biotypes.genus=biotyper.data.frame(as.data.frame(otu_table(with_biotype_genus)), k=3, manalysis=TRUE, distance.jsd=data.dist)

df_BET<-as.data.frame(harvest.biotypes.genus$BET$tab)
df_BET<-t(df_BET)
df_BET<-as.data.frame(df_BET)
df_BET <- tibble::rownames_to_column(df_BET, "Sample")

tax<-read.csv("read_genus.csv")
bet_scores_merge_tax<-merge(df_BET,tax,all.x=TRUE, all.y=TRUE)

bet_final<-bet_scores_merge_tax[,c(1:4,11)]
colnames(bet_final)<-c("ASV", "Biotype_1","Biotype_2","Biotype_3","Taxonomy")
```

Next we want to validate those taxa through a second method - here ALDEX2 based
```{r}
# data.cluster.aldex.clr.genus.harvest<-aldex.clr(as.data.frame(otu_table(with_biotype_genus)),
#                              sample_data(with_biotype_genus)$data.cluster,
#                              mc.samples=128,
#                              denom="all",
#                              verbose=TRUE,
#                              useMC=FALSE)

# glm.kru.harvest.data.cluster<-aldex.kw(data.cluster.aldex.clr.genus.harvest)

# glm.kru.harvest.data.cluster

# taxa_info <- data.frame(tax_table(with_biotype_genus))
# taxa_info <- taxa_info %>% rownames_to_column(var = "ASV")
# 
# sig_aldex2 <- glm.kru.harvest.data.cluster %>%
#   rownames_to_column(var = "ASV") %>%
#   filter(kw.eBH < 0.05) %>%
#   arrange(kw.eBH) %>%
#   dplyr::select(ASV, glm.eBH, kw.eBH)
# 
# sig_aldex2 <- left_join(sig_aldex2, taxa_info)
# 
# aldex_genera<-sig_aldex2[,10]
# 
# temp<-bet_final%>%filter(Taxonomy %in% aldex_genera)
# 
# final_df<-NULL
# 
# final_df$Taxonomy<-temp$Taxonomy
# final_df$Biotype_1<-temp$Biotype_1
# final_df$Biotype_2<-temp$Biotype_2
# final_df$Biotype_3<-temp$Biotype_3
# final_df<-as.data.frame(final_df)
# write.csv(final_df,"consensus taxa biotypeR genus.csv")
setwd("[...]/github pitcher plant addition ms resources")
final_df<-read.csv("consensus taxa biotypeR genus.csv")

Biotype_1_enriched<- final_df[,c(2:3)] %>% filter(Biotype_1>0)
Biotype_2_enriched<- final_df[,c(2,4)] %>% filter(Biotype_2>0)
Biotype_3_enriched<- final_df[,c(2,5)] %>% filter(Biotype_3>0)
```

*ASV diversity by genus-level presence in significant biotype clusters*

```{r}
setwd("[...]/github pitcher plant addition ms resources")
phy_object_final<-readRDS("final_pitcher_phyloseq")
with_biotype_genus<-readRDS("with_biotype_genus_harvest")
final_df<-read.csv("consensus taxa biotypeR genus.csv")
tax_csv<-read.csv("temp_tax_table.csv")

tax_rows_fixed<-data.frame(tax_csv, row.names = 1)
tax_matrix<-as.matrix(tax_rows_fixed)
TAX_16S<-tax_table(tax_matrix, errorIfNULL=TRUE)

seq_object<-phyloseq(otu_table(phy_object_final),TAX_16S,sample_data(phy_object_final),phy_tree(phy_object_final))

phylo<-prune_taxa(tax_table(seq_object)[,6] %in% final_df$Taxonomy,seq_object)

ASV_div_genus<-as.data.frame(colSums(otu_table(phylo)!=0))
ASV_div_genus<-tibble::rownames_to_column(ASV_div_genus, "ASV")
colnames(ASV_div_genus)<-c("ASV","Total ASVs")
meta<-as.data.frame(as.matrix((sample_data(with_biotype_genus))))
meta<-tibble::rownames_to_column(meta, "ASV")
meta<-meta[,-c(3:39)]

asv_div_df<-merge(ASV_div_genus,meta)

model<-aov(`Total ASVs`~ data.cluster,data=asv_div_df)
summary(model)

generate_label_df <- function(TUKEY, variable){

 #      Extract labels and factor levels from Tukey post-hoc
      Tukey.levels <- TUKEY[[variable]][,4]
      Tukey.labels <- data.frame(multcompLetters(Tukey.levels)['Letters'])

#      I need to put the labels in the same order as in the boxplot :
      Tukey.labels$treatment=rownames(Tukey.labels)
      Tukey.labels=Tukey.labels[order(Tukey.labels$treatment) , ]
      return(Tukey.labels)
}

# Tukey test to study each pair of treatment :
TUKEY <- TukeyHSD(x=model, 'data.cluster', conf.level=0.95)

library(multcompView)

LABELS <- generate_label_df(TUKEY,"data.cluster")

boxplot(asv_div_df$`Total ASVs`~asv_div_df$data.cluster)

x<-ggplot(asv_div_df,aes(x= data.cluster , y=`Total ASVs`)) + geom_boxplot(lwd=1, fatten = 1, alpha=1, fill= "grey") +
    # geom_point(position= position_jitterdodge(jitter.width = 0.3,jitter.height = 0,dodge.width = 0.75, seed = NA),shape = 21, alpha = 0.4, size = 8) +
    theme_cowplot()+
   theme(
  legend.position = 0,
  text=element_text(size=18),
  panel.grid.major = element_line(colour = "grey70", size = 0.1), panel.grid.minor = element_blank()) +  xlab("Biotype") + ylab("Total ASVs")
```

Just genera enriched in Bio1
```{r}
phylo<-prune_taxa(tax_table(seq_object)[,6] %in% Biotype_1_enriched$Taxonomy,seq_object)

ASV_div_genus<-as.data.frame(colSums(otu_table(phylo)!=0))
ASV_div_genus<-tibble::rownames_to_column(ASV_div_genus, "ASV")
colnames(ASV_div_genus)<-c("ASV","Total ASVs")
meta<-as.data.frame(as.matrix((sample_data(with_biotype_genus))))
meta<-tibble::rownames_to_column(meta, "ASV")
meta<-meta[,-c(3:39)]

asv_div_df<-merge(ASV_div_genus,meta)

model<-aov(`Total ASVs`~ data.cluster,data=asv_div_df)
summary(model)

#function from: https://r-graph-gallery.com/84-tukey-test.html

generate_label_df <- function(TUKEY, variable){

 #      Extract labels and factor levels from Tukey post-hoc
      Tukey.levels <- TUKEY[[variable]][,4]
      Tukey.labels <- data.frame(multcompLetters(Tukey.levels)['Letters'])

#      I need to put the labels in the same order as in the boxplot :
      Tukey.labels$treatment=rownames(Tukey.labels)
      Tukey.labels=Tukey.labels[order(Tukey.labels$treatment) , ]
      return(Tukey.labels)
}

# Tukey test to study each pair of treatment :
TUKEY <- TukeyHSD(x=model, 'data.cluster', conf.level=0.95)

LABELS <- generate_label_df(TUKEY,"data.cluster")

boxplot(asv_div_df$`Total ASVs`~asv_div_df$data.cluster)
y<-ggplot(asv_div_df,aes(x= data.cluster , y=`Total ASVs`)) + geom_boxplot(lwd=1, fatten = 1, alpha=1, fill= "grey") +
    # geom_point(position= position_jitterdodge(jitter.width = 0.3,jitter.height = 0,dodge.width = 0.75, seed = NA),shape = 21, alpha = 0.4, size = 8) +
    theme_cowplot()+
   theme(
  legend.position = 0,
  text=element_text(size=18),
  panel.grid.major = element_line(colour = "grey70", size = 0.1), panel.grid.minor = element_blank()) +  xlab("Biotype") + ylab("Total ASVs")

```







































